HardCodedStringCheckerSharp.UnitTests/CommandLineParserTest.cs

```diff
@@ -0,0 +1,166 @@
+
+         opts.Action.Should().Be( Action.ReportHCS );
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_FailOnHCSNotSpecified_SetsFailOnHCSToFalse()
+      {
+         var parser = new CommandLineParser();
+
+         string[] args = { @"C:\RepoRoot", "Report" };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.FailOnHCS.Should().BeFalse();
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_ThirdArgIsInvalid_ReturnsNull()
+      {
+         var parser = new CommandLineParser();
+
+         string[] args = { @"C:\RepoRoot", "Report", "Invalid" };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.Should().BeNull();
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_ThirdArgIsFailOnHCS_SetsFailOnHCSToTrue()
+      {
+         var parser = new CommandLineParser();
+
+         string[] args = { @"C:\RepoRoot", "Report", "--FailOnHCS" };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.FailOnHCS.Should().BeTrue();
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_ThirdArgIsExcludeButMissingExcludeFile_ReturnsNull()
+      {
+         var parser = new CommandLineParser();
+
+         string[] args = { @"C:\RepoRoot", "Report", "--Exclude" };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.Should().BeNull();
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_ThirdArgIsExcludeWithFourthArgExcludeFile_ReturnsExcludeFile()
+      {
+         var parser = new CommandLineParser();
+
+         string excludeFile = @"C:\Path\To\ExcludeFile.txt";
+         string[] args = { @"C:\RepoRoot", "Report", "--Exclude", excludeFile };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.ExcludeFile.Should().Be( excludeFile );
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_FourthArgIsInvalid_ReturnsNull()
+      {
+         var parser = new CommandLineParser();
+
+         string[] args = { @"C:\RepoRoot", "Report", "--FailOnHCS", "Invalid" };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.Should().BeNull();
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_FourthArgIsExcludeButMissingExcludeFile_ReturnsNull()
+      {
+         var parser = new CommandLineParser();
+
+         string[] args = { @"C:\RepoRoot", "Report", "--FailOnHCS", "--Exclude" };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.Should().BeNull();
+      }
+
+      [TestMethod]
+      public void ParseCommandLine_FourthArgIsExcludeWithFifthArgExcludeFile_ReturnsExcludeFile()
+      {
+         var parser = new CommandLineParser();
+
+         string excludeFile = @"C:\Path\To\ExcludeFile.txt";
+         string[] args = { @"C:\RepoRoot", "Report", "--FailOnHCS", "--Exclude", excludeFile };
+         var opts = parser.ParseCommandLine( args );
+
+         opts.ExcludeFile.Should().Be( excludeFile );
+      }
+   }
```

Comment:
:+1: for all these cases. I'm wondering if it might be worth adding one to make sure ordering doesn't matter? Or, if order _does_ matter, then having a test to enforce it would be nice too, IMO.

Reply:
With the tests I added, I tried to enforce all the rules, including ordering.

I'm trying to be backward compatible with the prior command line, so I don't have to update a bunch of TeamCity builds.  @bberton said this is used on Snagit and Camtasia, and maybe some Common builds too.

The first 2 command line parameters, `[RepoDirectory]` and `[Action]` are required.  These parameters are not named, so they must come before any optional parameters, and the order does matter.

The `[--FailOnHCS]` and `[--Exclude [ExcludeFile]]` parameters are optional.  You can supply none, one of them, or both.  Since they have --switch names, I can parse them out in any order, as long as they occur after the required parameters.

Given these rules, do you see any test cases I'm missing?

Reply:
Given those flags are optional and work in any order, I'm fine with what you have now 👍

Reply:
> I'm trying to be backward compatible with the prior command line, so I don't have to update a bunch of TeamCity builds

Awesome!  Though I'd say don't worry too much; make the changes you would like to have changed!  I'll update Snagit/Camtasia/Common builds if necessary.  Never be reluctant to make a change!  I'm just happy to see this getting some updates.  Looks good to me! :smiley: